src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,596 @@
+        {
+            lock.unlock();
+            return 0;
+        }
+        else
+            more_data_signal.wait(lock, [this, end_offset_] { return this->download_finished || this->current_offset >= end_offset_; });
+    }
+    LOG_TRACE(&Poco::Logger::get("RemoteCacheController"), "get more data to read");
+    lock.unlock();
+    return 0;
+}
+
+void RemoteCacheController::backgroupDownload(std::function<void(RemoteCacheController *)> const & finish_callback)
+{
+    download_thread = new ThreadPool(1);
+    auto task = [this, finish_callback]()
+    {
+        size_t n = 0;
+        size_t total_bytes = 0;
+        while (!remote_readbuffer->eof())
+        {
+            size_t bytes = remote_readbuffer->buffer().end() - remote_readbuffer->position();
+            out_file->write(remote_readbuffer->position(), bytes);
+            remote_readbuffer->position() += bytes;
+            total_bytes += bytes;
+            if (n++ % 10 == 0)
+            {
+                std::unique_lock lock(mutex);
+                current_offset += total_bytes;
+                total_bytes = 0;
+                flush();
+                lock.unlock();
+                more_data_signal.notify_all();
+            }
+        }
+        std::unique_lock lock(mutex);
+        current_offset += total_bytes;
+        download_finished = true;
+        flush();
+        out_file->close();
+        delete out_file;
+        out_file = nullptr;
+        remote_readbuffer = nullptr;
+        lock.unlock();
+        more_data_signal.notify_all();
+        finish_callback(this);
+        LOG_TRACE(
+            &Poco::Logger::get("RemoteCacheController"), "finish download.{} into {}. size:{} ", remote_path, local_path, current_offset);
+    };
+    download_thread->scheduleOrThrow(task);
+}
+
+void RemoteCacheController::flush()
+{
+    if (out_file != nullptr)
+    {
+        out_file->flush();
+    }
+    Poco::JSON::Object jobj;
+    jobj.set("schema", schema);
+    jobj.set("cluster", cluster);
+    jobj.set("remote_path", remote_path);
+    jobj.set("downloaded", download_finished ? "true" : "false");
+    jobj.set("last_mod_ts", last_mod_ts);
+    std::stringstream buf; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+    jobj.stringify(buf);
+
+    std::ofstream meta_file(local_path + "/meta.txt", std::ios::out);
+    meta_file << buf.str();
+    meta_file.close();
+}
+
+RemoteCacheController::~RemoteCacheController()
+{
+    delete out_file;
+
+    if (download_thread != nullptr)
+    {
+        download_thread->wait();
+        delete download_thread;
+    }
+}
+
+void RemoteCacheController::close()
+{
+    // delete the directory
+    LOG_TRACE(&Poco::Logger::get("RemoteCacheController"), "release local resource " + remote_path + ", " + local_path);
+    Poco::File file(local_path);
+    file.remove(true);
+}
+
+FILE * RemoteCacheController::allocFile(std::string * local_path_)
+{
+    if (download_finished)
+        *local_path_ = local_path + "/data.bin";
```

Comment:
It will be better to set `local_path_` in a separate method, this does not look good now.

Reply:
change as following
`std::tuple<FILE*, std::string> RemoteCacheController::allocFile()`